ports/esp32s2/common-hal/canio/CAN.c

```diff
@@ -0,0 +1,297 @@
+{
+    twai_status_info_t info;
+    twai_get_status_info(&info);
+    return info.tx_error_counter;
+}
+
+int common_hal_canio_can_receive_error_count_get(canio_can_obj_t *self)
+{
+    twai_status_info_t info;
+    twai_get_status_info(&info);
+    return info.rx_error_counter;
+}
+
+canio_bus_state_t common_hal_canio_can_state_get(canio_can_obj_t *self) {
+    twai_status_info_t info;
+    twai_get_status_info(&info);
+    if (info.state == TWAI_STATE_BUS_OFF || info.state == TWAI_STATE_RECOVERING) {
+        return BUS_STATE_OFF;
+    }
+    if (info.tx_error_counter > 127 || info.rx_error_counter > 127) {
+        return BUS_STATE_ERROR_PASSIVE;
+    }
+    if (info.tx_error_counter > 96 || info.rx_error_counter > 96) {
+        return BUS_STATE_ERROR_WARNING;
+    }
+    return BUS_STATE_ERROR_ACTIVE;
+}
+
+static void can_restart(void) {
+    twai_status_info_t info;
+    twai_get_status_info(&info);
+    if (info.state != TWAI_STATE_BUS_OFF) {
+        return;
+    }
+    twai_initiate_recovery();
+    // wait 100ms (hard coded for now) for bus to recover
+    uint64_t deadline = port_get_raw_ticks(NULL) + 100;
+    do {
+        twai_get_status_info(&info);
+    } while (port_get_raw_ticks(NULL) < deadline && (info.state == TWAI_STATE_BUS_OFF || info.state == TWAI_STATE_RECOVERING));
+}
+
+void canio_maybe_auto_restart(canio_can_obj_t *self) {
+    if (self->auto_restart) can_restart();
+}
+
+void common_hal_canio_can_restart(canio_can_obj_t *self) {
+    if (!common_hal_canio_can_auto_restart_get(self)) {
+        can_restart();
+    }
+}
+
+bool common_hal_canio_can_auto_restart_get(canio_can_obj_t *self) {
+    return self->auto_restart;
+}
+
+void common_hal_canio_can_auto_restart_set(canio_can_obj_t *self, bool value) {
+    self->auto_restart = value;
+    canio_maybe_auto_restart(self);
+}
+
+void common_hal_canio_can_send(canio_can_obj_t *self, mp_obj_t message_in)
+{
+    canio_maybe_auto_restart(self);
+    canio_message_obj_t *message = message_in;
+    bool rtr = message->base.type == &canio_remote_transmission_request_type;
+    twai_message_t message_out = {
+        .extd = message->extended,
+        .rtr = rtr,
+        .self = self->loopback,
+        .identifier = message->id,
+        .data_length_code = message->size,
+    };
+    if (!rtr) {
+        memcpy(message_out.data, message->data, message->size);
+    }
+    // Allow transmission to occur in background
+    twai_transmit(&message_out, 0);
+}
+
+bool common_hal_canio_can_silent_get(canio_can_obj_t *self) {
+    return self->silent;
+}
+
+bool common_hal_canio_can_deinited(canio_can_obj_t *self) {
+    return !self->tx_pin;
+}
+
+void common_hal_canio_can_check_for_deinit(canio_can_obj_t *self) {
+    if (common_hal_canio_can_deinited(self)) {
+        raise_deinited_error();
+    }
+}
+
+void common_hal_canio_can_deinit(canio_can_obj_t *self)
+{
+    if (self->tx_pin) {
+        (void)twai_stop();
```

Comment:
Why `(void)`?

Reply:
twai_stop returns an `esp_error_t`.  However, nothing useful can be done in the presence of an error.  For instance, we can't throw here.  While ignored return value diagnostics are apparently not enabled, casting the non-void return to void is a way to be explicit that we discard the return value.